packages/credential-provider-process/src/index.ts

```diff
@@ -0,0 +1,135 @@
+  ParsedIniData,
+  SharedConfigFiles,
+  SharedConfigInit
+} from "@aws-sdk/shared-ini-file-loader";
+import { exec } from "child_process";
+import {
+  getMasterProfileName,
+  parseKnownFiles
+} from "@aws-sdk/credential-provider-ini";
+
+const DEFAULT_PROFILE = "default";
+export const ENV_PROFILE = "AWS_PROFILE";
+
+export interface FromProcessInit extends SharedConfigInit {
+  /**
+   * The configuration profile to use.
+   */
+  profile?: string;
+
+  /**
+   * A promise that will be resolved with loaded and parsed credentials files.
+   * Used to avoid loading shared config files multiple times.
+   */
+  loadedConfig?: Promise<SharedConfigFiles>;
+}
+
+/**
+ * Creates a credential provider that will read from a credential_process specified
+ * in ini files.
+ */
+export function fromProcess(init: FromProcessInit = {}): CredentialProvider {
+  return () =>
+    parseKnownFiles(init).then(profiles =>
+      resolveProcessCredentials(getMasterProfileName(init), profiles, init)
+    );
+}
+
+async function resolveProcessCredentials(
+  profileName: string,
+  profiles: ParsedIniData,
+  options: FromProcessInit
+): Promise<Credentials> {
+  const profile = profiles[profileName];
+
+  if (profiles[profileName]) {
+    const credentialProcess = profile["credential_process"];
+    if (credentialProcess !== undefined) {
+      return await execPromise(credentialProcess)
+        .then((processResult: any) => {
+          let data;
+          try {
+            data = JSON.parse(processResult);
+          } catch {
+            throw Error(
+              `Profile ${profileName} credential_process returned invalid JSON.`
+            );
+          }
+
+          const {
+            Version: version,
+            AccessKeyId: accessKeyId,
+            SecretAccessKey: secretAccessKey,
+            SessionToken: sessionToken,
+            Expiration: expiration
+          } = data;
+
+          if (version !== 1) {
+            throw Error(
+              `Profile ${profileName} credential_process did not return Version 1.`
+            );
+          }
+
+          if (accessKeyId === undefined || secretAccessKey === undefined) {
+            throw Error(
+              `Profile ${profileName} credential_process returned invalid credentials.`
+            );
+          }
+
+          if (expiration) {
+            const currentTime = new Date();
+            const expireTime = new Date(expiration);
+            if (expireTime < currentTime) {
+              throw Error(
+                `Profile ${profileName} credential_process returned expired credentials.`
+              );
+            }
+            return {
+              accessKeyId,
+              secretAccessKey,
+              sessionToken,
+              expiration: Math.floor(new Date(expiration).valueOf() / 1000)
+            };
+          } else {
+            return {
+              accessKeyId,
+              secretAccessKey,
+              sessionToken
+            };
```

Comment:
Just one last comment: Do you really need the branching here? I believe the `expiration` is optional, so it's ok to be `undefined`. Just using one return will make the code cleaner